pkg/apis/pipeline/v1beta1/param_types.go

```diff
@@ -178,6 +178,47 @@ func validatePipelineParametersVariablesInTaskParameters(params []Param, prefix
 	return errs
 }
 
+func validatePipelineParametersVariablesInMatrixParameters(matrix []Param, prefix string, paramNames sets.String, arrayParamNames sets.String) (errs *apis.FieldError) {
+	for _, param := range matrix {
+		errs = errs.Also(param.validateArrayVariablesInMatrixParameter(prefix, paramNames, arrayParamNames))
+	}
+	return errs
+}
+
+func (p *Param) validateArrayVariablesInMatrixParameter(prefix string, paramNames sets.String, arrayParamNames sets.String) (errs *apis.FieldError) {
+	for idx, arrayElement := range p.Value.ArrayVal {
+		errs = errs.Also(validateArrayVariable(arrayElement, prefix, paramNames, arrayParamNames).ViaFieldIndex("value", idx).ViaFieldKey("matrix", p.Name))
+	}
+	return errs
+}
+
+func validateParametersInTaskMatrix(matrix []Param) (errs *apis.FieldError) {
+	for _, param := range matrix {
+		errs = errs.Also(param.validateMatrixParameterIsArrayTypeOnly())
```

Comment:
Small readability suggestion here:  given `param.validateMatrixParameterIsArrayTypeOnly` and `param.validateArrayVariablesInMatrixParameter` are each only used once I suggest in-lining them directly rather than having separate methods for them.

e.g.

```go
func validateParametersInTaskMatrix(matrix []Param) (errs *apis.FieldError) {
	for _, param := range matrix {
		if param.Value.Type != ParamTypeArray {
			errs = errs.Also(apis.ErrInvalidValue("parameters of type array only are allowed in matrix", "").ViaFieldKey("matrix", p.Name))
		}
	}
	return errs
}
```

The reason for this is that (at least for me, not sure what others think) method names like `validateArrayVariablesInMatrixParameter` are quite difficult to interpret while also keeping the control flow clear in my head.

Similarly for `validatePipelineParametersVariablesInMatrixParameters`, I think I'm able to more quickly understand what the func does when the `for` loops are nested:

```go
func validatePipelineParametersVariablesInMatrixParameters(matrix []Param, prefix string, paramNames sets.String, arrayParamNames sets.String) (errs *apis.FieldError) {
	for _, param := range matrix {
		for idx, arrayElement := range param.Value.ArrayVal {
			errs = errs.Also(validateArrayVariable(arrayElement, prefix, paramNames, arrayParamNames).ViaFieldIndex("value", idx).ViaFieldKey("matrix", p.Name))
		}
	}
	return errs
}
```

Not a big issue from my pov if you decide it's preferable as-is, but thought I'd mention it.

Reply:
moved the functions inline, and I agree that it's clearer that way! thank you @sbwsg :)